src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -29,17 +36,195 @@ case class NoOpFingerprint() {
+      val allFiles = ListBuffer[FileInfo]()
+      allFiles.appendAll(files.map(FileInfo(_)))
+
+      if (pathToDirectory.contains(dirPath)) {
+        // Map already contains this directory. Just append the files to its existing list.
+        pathToDirectory(dirPath).files.asInstanceOf[ListBuffer[FileInfo]].appendAll(allFiles)
+      } else {
+        var curDirPath = dirPath
+        // Create a new Directory object and add it to Map
+        val subDirs = ListBuffer[Directory]()
+        var directory = Directory(curDirPath.getName, files = allFiles, subDirs = subDirs)
+        pathToDirectory.put(curDirPath, directory)
+
+        // Keep creating parent Directory objects and add to the map if non-existing.
+        while (curDirPath.getParent != null && !pathToDirectory.contains(curDirPath.getParent)) {
+          curDirPath = curDirPath.getParent
+
+          directory = Directory(
+            if (curDirPath.isRoot) curDirPath.toString else curDirPath.getName,
+            subDirs = ListBuffer(directory),
+            files = ListBuffer[FileInfo]())
+
+          pathToDirectory.put(curDirPath, directory)
+        }
+
+        // Either root is reached (parent == null) or an existing directory is found. If it's the
+        // latter, add the newly created directory tree to its subDirs.
+        if (curDirPath.getParent != null) {
+          pathToDirectory(curDirPath.getParent).subDirs
+            .asInstanceOf[ListBuffer[Directory]]
+            .append(directory)
+        }
+      }
+    }
+
+    pathToDirectory(getRoot(files.head.getPath))
+  }
+
+  // Return file system root path from any path. E.g. "file:/C:/a/b/c" will have root "file:/C:/".
+  // For linux systems, this root will be "file:/". Other hdfs compatible file systems will have
+  // corresponding roots.
+  @tailrec
+  private def getRoot(path: Path): Path = {
+    if (path.isRoot) path else getRoot(path.getParent)
+  }
+
+  private def listLeafFiles(
+      path: Path,
+      pathFilter: PathFilter = PathUtils.DataPathFilter,
+      throwIfNotExists: Boolean = false,
+      fs: FileSystem): Seq[FileStatus] = {
+    try {
+      val (files, directories) = fs.listStatus(path).partition(_.isFile)
+      // TODO: explore fs.listFiles(recursive = true) for better performance of file listing.
+      files.filter(s => pathFilter.accept(s.getPath)) ++
+        directories.flatMap(d => listLeafFiles(d.getPath, pathFilter, throwIfNotExists, fs))
+    } catch {
+      case _: FileNotFoundException if !throwIfNotExists => Seq()
+      case e: Throwable => throw e
+    }
+  }
+}
+
+// modifiedTime is an Epoch time in milliseconds. (ms since 1970-01-01T00:00:00.000 UTC).
+case class FileInfo(name: String, size: Long, modifiedTime: Long)
+
+object FileInfo {
+  def apply(s: FileStatus): FileInfo = {
+    require(s.isFile, s"${FileInfo.getClass.getName} is applicable for files, not directories.")
+    FileInfo(s.getPath.getName, s.getLen, s.getModificationTime)
```

Comment:
@apoorvedave1 While resolving the conflict with my change, I found that `s.getPath.getName` discards the parent dir information. This might cause the loss of the location info unconsciously. What do you think?

Reply:
yeah you are correct. This was a conscious choice to remove the location info for removing duplication in the metadata.

Is this causing some problem somehow? Or do you foresee a problem in the future because of this? If so, we can do something like this: similar to `Content.files` api, we can expose a `def Content.fileInfos: Seq[FileInfo]` or similar api which adds location to file names. Alternately, we can discuss to keep full file paths in FileInfo object.

Please let me know your thoughts and we can explore either adding `def Content.fileInfos: Seq[FileInfo]` api or updating FileInfo object to hold full paths.

Reply:
I added `fileInfos()`  in my PR though it needs to be revised.
https://github.com/microsoft/hyperspace/pull/123/files#r485951598

And I used FileInfo in this way:
```
val curFileSet = location.allFiles
          .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
```

The point is, with this `apply` function, the parent dir info can be removed unintentionally - without informative function name.. or at least we need some comment here.

Reply:
oh i see, you are saying semantically it's not correct to allow both versions of FileInfo: one with just the name, another with full file path. Is my understanding correct?

I guess that's a fair point. Ideally we should have two different classes: One with just the name, One with full file path. E.g.
```scala
class FileInfo(filename: String, modifiedTime, size)

class FullPathFileInfo(path: Path, modifiedTime, size)
```
Please let me know if my understanding is correct. We can create an issue fix this limitation in another PR.

Reply:
How about we just store the full path in the log and just compress the index json file? Looks like we will always be building the full paths when hybrid scan is enabled.

Reply:
I agree that `FileInfo.name` should only store the final component of `Path` (same for `Directory.anem`) and `def apply(s: FileStatus): FileInfo` can be confusing when the parent directory is stripped away. Looks like we don't really need this helper function and be explicit about the behavior - i.e., call constructor? (used once in code, and many in tests, but you can define a helper function in `IndexLogEntryTest`.

Reply:
With `lazy val` instead of `def`, the performance seems okay once `allFileInfos` is built.
On second thought, as `FileInfo` is defined under `Content` which is tree based structure, it's reasonable to keep only filename.

I need `FileInfo( full path string, size, modification time)` to compare & intersect 2 lists of files. I think it's okay to reuse the current `FileInfo` for hybrid scan as it's just a simple case class and we could revisit the structure & api later.